streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KTable.scala

```diff
@@ -245,24 +233,11 @@ class KTable[K, V](val inner: KTableJ[K, V]) {
    * one for each matched record-pair with the same key
    * @see `org.apache.kafka.streams.kstream.KTable#join`
    */
-  def join[VO, VR](other: KTable[K, VO])(
-    joiner: (V, VO) => VR,
-    materialized: Materialized[K, VR, ByteArrayKeyValueStore]
+  def join[VO, VR](other: KTable[K, VO])(joiner: (V, VO) => VR)(
+    implicit materialized: Materialized[K, VR, ByteArrayKeyValueStore]
```

Comment:
Sorry if I'm being dense, but is it ok for Materialized to be implicit?

It seems to me that this would allow us to elide the parameter as long as there is a Materialized of the correct generic type in context, right?

But the Materialized builders aren't freely interchangeable just based on their generic types. In particular, if they have a name, then it would be incorrect to give the same name to every store in the topology. At first glance this seems like more room for error than I'm comfortable with for implicits.

Reply:
You're not being dense.
The default `Materialized` given as implicit:
https://github.com/apache/kafka/blob/bb4cc4962894cf930c6b1cc01b121444f4ebb106/streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/ImplicitConversions.scala#L76
Doesn't specify any name or other configuration for the store. Therfore it would behave the same as if we didn't give one.
In fact we are doing this for `aggregate` and other transformations.

Reply:
Ok, I agree that's safe. I just had another thought reviewing this, though...

It's not immediately obvious that this would be the case, but join results are not materialized by default. Each side of the join must maintain the state of its stream, but since the final result can always be computed from the left and right stores, we don't need to materialize the join result.

So if I happen to have this implicit in scope, then all the joins in my topology will become materialized. Obviously, this will have some impact on resource usage.

I think that I actually have no way to specify that I *don't* want a join to be materialized in the current API.

Note that most of the other KTable operators are similar in behavior. For example: `table.filter.filter.mapValues.toStream` with no materializations actually only needs one state store (for the original table), but if we implicitly materialized them all, you'd wind up storing the data 4 times.

Reply:
Now, arguably this is a bug in the Java code, since AFAICT it's pointless to materialize the join if there's no queriable name.

Looking in `KTableImpl`, the join does this:
```
// only materialize if specified in Materialized
        if (materializedInternal != null) {
            kTableJoinNodeBuilder.withMaterializedInternal(materializedInternal);
        }
```

But the filter, for example, does this:
```
// only materialize if the state store is queryable
        final boolean shouldMaterialize = materializedInternal != null && materializedInternal.isQueryable();
```

Maybe we should go ahead and fix this logic, and then the implicit would be fine.

But I'm still not sure that it's great to require Materialized as an argument, effectively requiring the implicit as well, or a bunch of dummy arguments in my topology, just so they can be ignored later.

It's a bummer, because needing to support two variants means we can't do the currying approach at all, which means we can't benefit from the improved type inference.

Maybe the way out is just to have `join` and `materializedJoin` (and `innerJoin` and `materializedInnerJoin`, etc.). I don't know...

Reply:
@guozhangwang WDYT? Is there some benefit to always requiring (possibly anonymous) Materialized on joins?

I could see it being advantageous to make the serdes available for possible optimizations, but I don't know if this is the right mechanism for that anyway. I.e., you have to pass in something called "materialized", but it actually doesn't materialize anything; just registers serdes -- this seems confusing.

Reply:
@vvcephei The only places where we've expected an `implicit materialized: Materialized` are in `count()`s and in `aggregate()`s, so I don't see why in your example `table.filter.filter.mapValues.toStream` it would materialize 4 times?

Reply:
Sorry, my comment was ambiguous. I was just saying that such would be the result *if* we make materialized implicit across the board in `KTable`.